### ISSUE TITLE
`replenishReserves` is subject to be bricked for some conditions

### ISSUE THREAD

**Issue by c4-submissions** [code-423n4/2023-09-maia-findings#689]
# Lines of code

https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchPort.sol#L181
https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchPort.sol#L214

# Vulnerability details

## Impact
User funds will remain stuck at the strategy contract
## Proof of Concept
`BranchPort` contracts are used to keep user funds. They also have the idea of earning yields with some added strategies. 
Accordingly, if the underlying token is approved and added as a strategy token, it can be transferred to the strategy contract to earn some rewards for the amount transferred.

The transfer of the funds is managed by the [manage](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchPort.sol#L144) function as  below;
```solidity
Contract: BranchPort.sol

144:     function manage(address _token, uint256 _amount) external override requiresPortStrategy(_token) {
145:         // Cache Strategy Token Global Debt
146:         uint256 _strategyTokenDebt = getStrategyTokenDebt[_token];
147: 
148:         // Check if request would surpass the tokens minimum reserves
149:         if (_amount > _excessReserves(_strategyTokenDebt, _token)) revert InsufficientReserves();
150: 
151:         // Check if request would surpass the Port Strategy's daily limit
152:         _checkTimeLimit(_token, _amount);
153: 
154:         // Update Strategy Token Global Debt
155:         getStrategyTokenDebt[_token] = _strategyTokenDebt + _amount;
156:         // Update Port Strategy Token Debt
157:         getPortStrategyTokenDebt[msg.sender][_token] += _amount;
158: 
159:         // Transfer tokens to Port Strategy for management
160:         _token.safeTransfer(msg.sender, _amount);
161: 
162:         // Emit DebtCreated event
163:         emit DebtCreated(msg.sender, _token, _amount);
164:     }
```

As can be seen on [Line:155](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchPort.sol#L155) and [Line:157](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchPort.sol#L157), the transferred amount is recorded as a debt to be paid later by the approved address.

The repaying is called as replenishing reserves and implemented by `replenishReserves` functions. While both `replenishReserves` functions ([1](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchPort.sol#L167),[2](https://github.com/code-423n4/2023-09-maia/blob/f5ba4de628836b2a29f9b5fff59499690008c463/src/BranchPort.sol#L188)) serve the same, both have a common requirement to ensure the debt is paid. It's by checking the balance of the RootPort contract before and after the transfer.

```solidity
Contract: BranchPort.sol

180:         // Check if _token balance has increased by _amount
181:         require(ERC20(_token).balanceOf(address(this)) - currBalance == _amount, "Port Strategy Withdraw Failed");
```
and
```solidity
Contract: BranchPort.sol

212:         // Check if _token balance has increased by _amount
213:         require(
214:             ERC20(_token).balanceOf(address(this)) - currBalance == amountToWithdraw, "Port Strategy Withdraw Failed"
215:         );
```

The strategy tokens are chosen wisely and currently implemented by the contract owners. They don't include any fee on transfer or rebasing tokens but mostly solid strategies like keeping the stables and native chain tokens.

However, it should be assumed that many tokens, especially the stables tokens, are not decentralized and subject to be upgraded or having an updated setting. This includes implementing fees. 

With a solid example we can check the [USDT](https://coinmarketcap.com/currencies/tether/) token which is the 3rd biggest token in marketcap and mos preferred stable token accordingly.

And [here](https://etherscan.io/token/0xdac17f958d2ee523a2206206994597c13d831ec7#writeContract#F8)'s their contract on Etherscan. 
And [here](https://etherscan.deth.net/address/0xdac17f958d2ee523a2206206994597c13d831ec7) is the human readable version of it.

Accordingly it can be seen that they have a [setParam](https://etherscan.io/token/0xdac17f958d2ee523a2206206994597c13d831ec7#writeContract#F8)
function which is a setter for the fees. 

![setParam](https://user-images.githubusercontent.com/65364747/273001743-766958e6-66ee-453b-b57d-9aabf12221e0.png)

The [function](https://etherscan.deth.net/address/0xdac17f958d2ee523a2206206994597c13d831ec7) is below;
```solidity
Contract: TetherToken.sol

425:    function setParams(uint newBasisPoints, uint newMaxFee) public onlyOwner {
426:        // Ensure transparency by hardcoding limit beyond which fees can never be added
427:        require(newBasisPoints < 20);
428:        require(newMaxFee < 50);
429:
430:        basisPointsRate = newBasisPoints;
431:        maximumFee = newMaxFee.mul(10**decimals);
432:
433:        Params(basisPointsRate, maximumFee);
434:    }
```
While the implemented `maximumFee` is zero now, this might change any time up to the centralized entity's strategies.
![maxFee](https://user-images.githubusercontent.com/65364747/273002434-7ac47897-0613-40c9-915d-5aa84c5d7d2c.png)

Also, since the USDT token is still using their non-upgradable contract, it's reminded that they can anytime deploy another contract in a proxy form and make the currency a double-entry point token like the [True USD](https://coinmarketcap.com/currencies/trueusd/) (Global Rank: 22) had experienced which may create more troubles to the strategy. Compound had a critical vulnerability in the past for this reason with an expected damage of 80M $ if exploited. More info can be found [here](https://blog.openzeppelin.com/compound-tusd-integration-issue-retrospective).

The other largest stable is [USDC](https://coinmarketcap.com/currencies/usd-coin/) and they're currently sitting on the 6th rank as per the martketcap. 
[Here](https://etherscan.io/token/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48#code)'s their contract. 
As also can be seen that they're implementing a proxy contract which means that their implementation can change without any notice and they can start implementing fees too as well.

In conclusion, there are fees implemented in the largest stable token's flow, and if they're set greater than zero, the `replenishReserves` will revert on above lines (181 & 214) due to the `amountToWithdraw` will be cut by the strategy token as per their contract and only the amount of `(amountToWithdraw - fee)` will be transferred to the `BranchPort` address. And the require statements will revert the transactions bricking the debt payment. This will result the used funds to remain at strategy contracts.

## Tools Used
Manual Review

## Recommended Mitigation Steps
We recommend not implementing the before/after balance checks to ensure the transfers. Instead implement safeTransfer libraries to bubble up any error that might occur.

## Assessed type

Token-Transfer
